File: frontend/src/api/activity.ts
import { BaseURL } from "./base.ts";
import type { Category } from "./content.ts";

export type Activity = {
    category: Category;
    channelId?: string;
    id?: number;
    status: ActivityStatus;
    videoId?: string;
};

export enum ActivityStatus {
    blockChannel = "blockChannel",
    completed = "completed",
}

export const createActivity = async (activity: Activity): Promise<Activity> => {
    const resp = await fetch(`${BaseURL}/api/activity`, {
        body: JSON.stringify(activity),
        headers: { "Content-Type": "application/json" },
        method: "POST",
    });

    if (!resp.ok) {
        throw new Error("Failed to create activity");
    }

    return resp.json();
};

File: app/user/activity.go
package user

import "content-oracle/app/database"

type Activity struct {
	blockedVideoRepository   *database.BlockedVideoRepository
	blockedChannelRepository *database.BlockedChannelRepository
}

func NewActivity(
	blockedVideoRepository *database.BlockedVideoRepository,
	blockedChannelRepository *database.BlockedChannelRepository,
) *Activity {
	return &Activity{
		blockedChannelRepository: blockedChannelRepository,
		blockedVideoRepository:   blockedVideoRepository,
	}
}

func (s *Activity) BlockChannel(channelID, status string) (*database.BlockedChannel, error) {
	return s.blockedChannelRepository.Create(database.BlockedChannel{
		ChannelID: channelID,
		Status:    status,
	})
}

func (s *Activity) BlockVideo(videoID, status string) (*database.BlockedVideo, error) {
	return s.blockedVideoRepository.Create(database.BlockedVideo{
		VideoID: videoID,
		Status:  status,
	})
}
